### ISSUE TITLE
ICE: `Unimplemented` when trying to select an implementation for constant trait item reference

### ISSUE THREAD

**Issue by petrochenkov** [rust-lang/rust#29595]
Code:

```
#![feature(associated_consts)]

trait Tr {
    const C: Self;
}

fn main() {
    let a: u8 = Tr::C;
}
```

Output:

```
<anon>:4:5: 4:19 error: internal compiler error: Encountered error `Unimplemented` when trying to select an implementation for constant trait item reference.
<anon>:4     const C: Self;
             ^~~~~~~~~~~~~~
note: the compiler unexpectedly panicked. this is a bug.
note: we would appreciate a bug report: https://github.com/rust-lang/rust/blob/master/CONTRIBUTING.md#bug-reports
note: run with `RUST_BACKTRACE=1` for a backtrace
thread 'rustc' panicked at 'Box<Any>', ../src/libsyntax/diagnostic.rs:176
```

Reproduces on nightly.

**Comment by samdoiron** [comment]
Triage: Able to reproduce on `rustc 1.7.0-nightly (abf60b20a 2015-12-19)`

**Comment by petrochenkov** [comment]
Fixed in 1.17.0, needs test.
```
rustc 1.17.0-nightly (306035c21 2017-02-18)
error[E0277]: the trait bound `u8: Tr` is not satisfied
 --> <anon>:8:17
  |
8 |     let a: u8 = Tr::C;
  |                 ^^^^^ the trait `Tr` is not implemented for `u8`
  |
  = note: required by `Tr::C`

error: aborting due to previous error
```
